dds_glossary/neo.py

```diff
@@ -0,0 +1,167 @@
+
+    logging.info("Importing concept -> concept 'broader' relationships...")
+
+    # Load concept "broader" their relationships
+    for semantic_relation in semantic_relations:
+        edge_type = semantic_relations[0]["type"]
+
+        edge =  ("Concept", semantic_relation["source_concept_iri"], "Concept", semantic_relation["target_concept_iri"])
+        
+        query, args = build_edges_import_query_and_args([edge_type], [edge])
+        driver.execute_query(query, args)
+
+    logging.info("Imported concept -> concept 'broader' relationships")
+
+
+def build_nodes_import_query_and_args(labels: List[str], nodes: List[Dict[str, Any]]):
+    """
+    Bulk import nodes into Neo4J
+
+    ## Example
+
+    > build_nodes_import_query_and_args(["Hello", "World"], [{"a": 1, "b": 2}, {"a": 1, "c": 10}])
+    (
+        "MERGE (e_0:Hello:World) {a: $a_0, b: $b_0}\nMERGE (e_1:Hello:World) {a: $a_1, c: $c_1}",
+        {'a_0': 1, 'b_0': 2, 'a_1': 1, 'c_1': 10}
+    )
+    """
+    query_args = {}
+    for idx, node in enumerate(nodes):
+        for k, v in node.items():
+            query_args[f"{k}_{idx}"] = v
+
+    schema_keys = set()
+    for node in nodes:
+        for k in node.keys():
+            schema_keys.add(k)
+
+    node_labels_str = ':'.join(labels)
+    
+    query_rows = []
+    for idx, node in enumerate(nodes):
+        schema_kv = [f"{k}: ${k}_{idx}" for k in node.keys()]
+        query_row = f"MERGE (e_{idx}:{node_labels_str} {{{', '.join(schema_kv)}}})"
+        query_rows.append(query_row)
+
+    query = "\n".join(query_rows)
+    return query, query_args
+
+
+def build_edges_import_query_and_args(labels: List[str], edges: List[Tuple[str, str, str, str]]):
+    """
+    Bulk import nodes into Neo4J
+
+    ## Example
+
+    > build_edges_import_query_and_args(["IsFrom"], [("Concept", "def", "ConceptScheme", "abcd")])
+    (
+        "MATCH (src_0:Concept {iri: $iri_src_0}), (tgt_0: ConceptScheme {iri: $iri_tgt_0})\nWITH src_0, tgt_0\nMERGE (src_0)-[r_0:IsFrom]->(tgt_0)",
+        {'iri_src_0': 'def', 'iri_tgt_0': 'abcd'}
+    )
+    """
+    query_args = {}
+    for idx, edge in enumerate(edges):
+        _, source_iri, _, target_iri = edge 
+        query_args[f"iri_src_{idx}"] = source_iri
+        query_args[f"iri_tgt_{idx}"] = target_iri
+    
+    edge_labels_str = ":".join(labels)
+
+    matches = []
+    withs = []
+    merges = []
+    for idx, edge in enumerate(edges):
+        source_label, source_iri, target_label, target_iri = edge 
+        matches.extend([
+            f"(src_{idx}:{source_label} {{iri: $iri_src_{idx}}})",
+            f"(tgt_{idx}:{target_label} {{iri: $iri_tgt_{idx}}})"
+        ])
+        withs.extend([f"src_{idx}", f"tgt_{idx}"])
+        merges.extend([f"(src_{idx})-[r_{idx}:{edge_labels_str}]->(tgt_{idx})"])
+
+    query = f"""
+    MATCH {', '.join(matches)}
+    """
+    for merge in merges:
+        query += f"MERGE {merge}"
```

Comment:
For future reference:
https://neo4j.com/docs/apoc/current/overview/apoc.import/
https://neo4j.com/labs/apoc/4.2/overview/apoc.import/apoc.import.json/